Suggested changes

bradleyharden opened the pull request:
Here's what I managed to implement today. It's not everything mentioned in my comment, but the major things are there, I think.

I tried to make the commits atomic. They aren't perfect, but it should be easier to review them individually than all at once.

In particular, I overhauled the `rtc` module. Take a look and see what you think. I was able to remove the `Output1k` and `Output32k` type parameters entirely from `OscUlp32k` and `Xosc32k`.

bradleyharden committed:
clock/v2: Migrate counting traits to the typelevel module

Move the `Counter`, `Increment`, `Decrement`, etc. traits to the
typelevel module. They are purely type-level tools that are not specific
to the `clock` module. However, the `Enabled` type is intimately tied to
the `clock` module, so it remains.

bradleyharden committed:
clock/v2: Rename identity types for consistency

Several pairs of existing types in the `clock` module have the same
relationship as `PinId` and `DynPinId`. Rename these types to show the
symmetry.

Also rename all `marker::*` types to end with an `Id` suffix and lift
them out of the `marker` modules. Add documentation to explain that
these types represent the clock *identity*, not the clock itself. This
is simpler and easier to understand than the existing `marker` approach.
Rename the `*SourceMarker` traits to `*SourceId` as well.

bradleyharden committed:
clock/v2: Move `Enabled` from types module to top-level

Now that `Counter` and its associated traits have been moved to the
`typelevel` module, the `types` module only contains the `Enabled`
wrapper type. Migrate this to the top-level of the `clock::v2` module,
where it fits better. Leave the empty `types` module for now, since it
contains documentation we may want to salvage in the future.

bradleyharden committed:
clock/v2: Consolidate `*Source` traits into single `Driver` trait

All of the existing `*Source` traits are effectively identical,
differing only by the trait bound applied to the associated type.
Because that trait bound must be repeated at each usage site, keeping
these as separate traits serves no useful purpose. Instead, combine
them all into a single trait. Name it `Driver` to distinguish the
direction of the relationship. `Driver` represents the forward
direction, where a clock acts as a source for another clock, whereas
`Source` represents the reverse direction, where a clock receives input
from another clock.

Neglect updating the `rtc` module for now, because of future, planned
refactoring.

bradleyharden committed:
Rtc: Overhaul `rtc` module to simplify the design

Rather than wrap the `crate::rtc::Rtc` struct, instead create an
`RtcOsc` struct that mimics a `Pclk` for the RTC. A future revision of
the `crate::rtc` module can require the `RtcOsc` as proof of proper
clock configuration. This structure relieves the OscUlp32k and Xosc32k
types from their `Output1k` and `Output32k` type parameters, which
greatly simplifies the design.

bradleyharden committed:
Xosc: Simplify the `xosc` module

Modify the `Mode` trait so that the type parameter can be removed from
`CrystalMode`. This eliminates redundancy when specifying the full type
of an `Xosc`. Also add a `configure` function to the `Mode` trait, so
that the `Xosc::enable` function can be generic. Rename the variants of
`CrystalCurrent` to remove the frequency range. Add this information to
the documentation, and refer to the datasheet for optimizing power
usage. Adjust the way a frequency, in `Hertz`, is converted to a
`CrystalCurrent`.

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc32k.rs:
```diff
@@ -200,28 +190,19 @@ impl Sealed for CrystalMode {}
 /// - An output state of a 32 kHz signal ([`Active32k`]/[`Inactive32k`])
 /// - An output state of a 1 kHz signal ([`Active1k`]/[`Inactive1k`])
-pub struct Xosc32k<M, X, Y>
+pub struct Xosc32k<M>
```
`Active32k` and `Active1k` has to come back in some form. I guess this is what you're working on right after the conversation about Gclks also depending on `EN32K` switch.

bradleyharden replied:
Yes, I didn't recognize this until later. I'm still playing with alternative ideas. Once I'm out of ideas, I'll see if you like any of them.

bradleyharden replied:
See the latest approach just pushed

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -26,169 +26,125 @@ use typenum::U0;
-            freq @ 8_000_001..=16_000_000 => (CrystalCurrent::LowFreq8mTo16m, freq),
-            freq @ 16_000_001..=24_000_000 => (CrystalCurrent::MedFreq16mTo24m, freq),
-            freq @ 24_000_001..=48_000_000 => (CrystalCurrent::HighFreq24mTo48m, freq),
-            _ => return Err(()),
```
The most important property of `CrystalConfig` was that it was instantiated externally to the `Xosc::new` so if user failed to create it, it was impossible to instantiate `Xosc` itself. "Missing ingredient" approach. I found it more elegant and _composable_ than `panic`.

bradleyharden replied:
~~Huh? The [existing implementation](https://github.com/vccggorski/atsamd/blob/a7378f5f5a28c84c028a7e1f5fca2c5f8295b454/hal/src/thumbv7em/clock/v2/xosc.rs#L131-L140) converts the `src_freq` into a `CrystalCurrent` inside the `new` function, returning an `Err(())` if it's out of bounds.~~

Edit: Sorry, I misread the `impl` block there. You are correct. I think my argument below still stands though.

I agree that `TryFrom<Hertz>` or some other fallible API is better in the general case. But in this specific case, I don't think that makes sense. Consider the reasons it might be out of bounds to begin with:
- The frequency is statically known and valid, but  the user made a typo.
    - I think `panic!` is the best choice here.
- The frequency is statically known and truly outside the 8-48 MHz range.
    - I don't think this is supported by the XOSC. Again, `panic!` makes sense.
- The frequency is not statically known and can be changed.
    - I suppose this is possible, but I've never seen or heard of a dynamically adjustable crystal oscillator. Have you?

I guess my question is this: What can a user *do* with an `Err` other than `.unwrap()` it? If every reasonable path leads to `.unwrap()`, then why bother returning a `Result` in the first place?

glaeqen replied:
Ok then

bradleyharden replied:
One other thought. Maybe we should just force the user to specify *both* the frequency and current. That way there's no possibility of error or panic. And we can repeat the datasheet documentation on the `CrystalCurrent` struct, so users don't have to go digging in the datasheet. Do you like that better?

bradleyharden replied:
The more I think about it, the more I like forcing the user to make a choice. We can auto-select a `CrystalCurrent` based on the frequency, but there are quite a few situations where the choice is not optimal or undefined. In those situations, I think it makes sense to push the decision to the user.

The `from_crystal` function would take both a frequency and a `CrystalCurrent`. Thoughts?

bradleyharden replied:
I implemented this change. Now, the user is forced to manually choose a current in `from_crystal`. The documentation for `CrystalCurrent` provides a summary of the datasheet. Does that work?

vcchtjader replied:
I support this idea that the user must make a conscious decision for both frequency and current.
As noted they are not set in stone (frequency - current relation) and with the datasheet documentation provided through docs users easily can find what to pick.

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -26,169 +26,125 @@ use typenum::U0;
+#[allow(missing_docs)]
 pub enum CrystalCurrent {
-    /// 8MHz
-    BaseFreq8m,
```
@vcchtjader will probably have more insight here but I also have some doubts about replacing variant names with vague "highs" and "lows". But maybe that's fine :thinking:

bradleyharden replied:
I'm not married to the names. But I wanted to avoid something like this:
```rust
let config = CrystalConfig::new(25.mhz());
let xosc = Xosc::from_crystal(token, xin, xout, config);
    .set_current(CrystalCurrent::MedFreq16mTo24m)
    .enable();
```
This code is perfectly reasonable and valid for a 25 MHz oscillator with a low capacitive load. Yet the code seems to directly contradict itself. In one line, it says the frequency is 25 MHz. Then in another, it says the frequency is in the range 16 MHz to 24 MHz.

If some other user came along later and read this code, I think it would seem like a bug. At the very least, it's confusing if you don't already know the datasheet well.

I don't particularly care what the variant names are. But I would prefer them not to contain frequencies, because the frequencies are only *suggestions* by the datasheet.

glaeqen replied:
Ok, makes sense

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -311,37 +267,83 @@ impl<X: XoscNum> XoscToken<X> {
+    pub trait Mode<X: XoscId>: Sealed {
+        const FROM_CRYSTAL: bool;
+        type Config;
+        fn configure(token: &mut XoscToken<X>, config: &Self::Config);
```
Instead of this `private` idiom, I consider approach I took in `Dfll` module slightly nicer. `fn configure` should take `&self` so one needs an instance of a mode that is inconstructible by a user code (therefore public nature of trait based methods is irrelevant). In `dfll` module this function is called `fn enable` which IMO fits nicely with a notion of _mode specific enable procedure_. This becomes apparent in `Dfll::enable` (`self.mode.enable(&mut self.token);`). Quite clean.

bradleyharden replied:
Hmm. That's another approach. But in that case, shouldn't the corresponding trait be `unsafe` to implement? That way anyone editing the code will know that implementing types should **not** be constructable by users.

There's also another problem with that approach. In the current implementation, specifying an `Xosc` in `CrystalMode` requires redundant types, i.e. `Xosc<XoscId0, CrystalMode<XoscId0>>`. You have to specify `XoscId0` *twice*. I refactored the `Mode` trait to get rid of that redundancy. That's why I added the type parameter and associated `Config` type. With the new implementation, you can just write `Xosc<XoscId0, CrystalMode>`. I don't think you can get rid of that redundancy with the `Self` approach. In that case, the `CrystalMode` type would have to be generic.

However, I think your approach could still work if I rearrange things slightly. I just have to make `CrystalConfig` public, but make it impossible to construct by users.

Although, on second thought, that might not work for `type Config = ();`. I would probably have to create some dummy type that can be instantiated by the `xosc` module but not by users. I'll have to think more about this.

glaeqen replied:
Maybe `Xosc<CrystalMode<XoscId0>>` and `Xosc<ClockMode<XoscId0>>` with associated `type Id: XoscId` constant for `Mode` trait is the way?
```rust
pub trait Mode: Sealed {
    type Id: XoscId;
    fn configure(&self, token: &mut XoscToken<Self::Id>);
}
pub struct ClockMode<X: XoscId>(PhantomData<X>)
impl<X: XoscId> Mode for ClockMode<X> {
    type Id: XoscId = X;
    fn configure(&self, token: &mut XoscToken<Self::Id>) {
    ....
    }
}
pub struct CrystalMode<X: XoscId>{
    xout: XOut<X>,
    current: CrystalCurrent,
    amplitude_loop_control: bool,
    low_buf_gain: bool,
}
impl<X: XoscId> Mode for CrystalMode<X> {
    type Id: XoscId = X;
    fn configure(&self, token: &mut XoscToken<Self::Id>) {
    ....
    }
}
```

Then you have access to `Id` everywhere you need it, including `Xosc` itself through `Mode::Id`. And notion of `Config` can be then flattened back to `Mode`. So all the fields `CrystalConfig` is holding can be moved back to `CrystalMode`. Maybe `Mode` should `unsafe` to implement with the comment that structs implementing it must be inconstructible in user code.

bradleyharden replied:
Let me back up for a moment and explain in more detail why I was trying to refactor `xosc` (and `dpll`) in the first place.

Here is a list of the existing clock types and their generic parameters:
- `AhbClk<A> where A: AhbId`
- `ApbClk<A> where A: ApbId`
- `Gclk<G, I> where G: GclkId, I: GclkSourceId`
- `GclkIn<G, I> where G: GclkId, I: GclkIo<G>`
- `GclkOut<G, I> where G: GclkId, I: GclkIo<G>`
- `Pclk<P, I> where P: PclkId, I: PclkSourceId`
- `Dfll<M> where M: Mode`
- `Dpll<D, M> where D: DpllId, M: SrcMode<D>`
- `Xosc<X, M> where X: XoscId, M: Mode`
- `OscUlpBase`
- `OscUlp1k`
- `OscUlp32k`
- `XoscBase<M> where M: Mode`
- `Xosc32k`
- `Xosc1k`
- `RtcOsc<I> where I: RtcSourceId`

The pattern here is clear:
- Where a clock has more than one instance (e.g. `Gclk`), the first type parameter is an identifier for the specific instance (e.g. `Gclk5Id`). If there is only one instance (e.g. `Dfll`), there is no corresponding type parameter.
    - Note, this lets us easily create type aliases of the form `pub type Gclk5<I> = Gclk<Gclk5Id, I>;`
- Where a clock needs some amount of type-level configuration, it takes a second type parameter, usually indicating something about the source or output of the clock.

In general, the second type parameter is populated by a *single* type that *does not itself* have any type parameters. In fact, that was one of the primary reasons to create the `Id` types in the first place. However, that pattern is currently broken in two places:
- The second type parameter for `Xosc<X, M>` can be `CrystalMode<X> where X: XoscId`
- The second type parameter for `Dpll<D, M>` is one of three `*Driven` structs that each takes two type parameters. The first of each `*Driven` struct is always `D: DpllId`, and the second is one of `PclkSourceId == GclkId`, `Xosc32kId`, or `X: XoscId`.

Notice how the concrete types used for the second type parameter always *repeat* the first type parameter, i.e. `D` for `Dpll` and `X` for `Xosc`. This is redundant, and it breaks the otherwise consistent pattern throughout the rest of the module.

My aim for refactoring both the `xosc` and `dpll` modules was to remove the redundancy and make the entire `clock` module uniform, where the second type parameter is always a *single* type with no type parameters of its own.

Your latest proposal would actual break the pattern further. It would define `Xosc` as `Xosc<M> where M: Mode<X> where X: XoscId`. If we took that approach, we would no longer be able to define `Xosc0` and `Xosc1` aliases.

Moreover, if we want a consistent pattern, we need to *remove* the type parameter from `CrystalMode`, pushing it onto the trait, i.e. `Mode<X>`. That's what my initial proposal does.

I agree that the first draft of my implementation was not as clear as it could be, but I think that can be improved. I'll try to come up with an implementation you like better. But I would really like to make the pattern consistent across all of the clocks. I think that will help tremendously withe comprehensibility. Eventually, you should be able to write:
- `Dpll0<Gclk5Id>` for a `Pclk` driven DPLL
- `Dpll0<Xosc32kId>` for an XOSC32K driven DPLL
- `Dpll0<Xosc1Id>` for an XOSC driven DPLL
- `Xosc1<CrystalMode>` for a crystal driven XOSC

I think this structure is really uniform and worth bending the implementation to accommodate.

bradleyharden replied:
I've now implemented all of these changes. I think there is only one situation where I couldn't remove a type parameter from the configuration type. `dfll::OpenLoop` can be a single type, without type parameters. But `dfll::ClosedLoop<G>` *must* have a type parameter, because it parameterizes which `Pclk<DfllId, G>` was used. The `G` type parameter only applies in `ClosedLoop` `Mode`, so that is the natural way to model it.

This case is different than the `Xosc` and `Dpll` cases above, because the `G` type parameter is not redundant, like the `X` and `D` type parameters were.

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -311,37 +267,83 @@ impl<X: XoscNum> XoscToken<X> {
-/// Struct representing a clock mode for [`Xosc`]
+    /// Private trait to restrict access to the `configure` function
+    pub trait Mode<X: XoscId>: Sealed {
+        const FROM_CRYSTAL: bool;
```
I don't like `bool`s like this, always feel like a code smell to me. I'd prefer actual call to `token.from_crystal/from_clock` in respective `fn configure` implementations.

bradleyharden replied:
Well, I wanted to preserve the order of calls within `.enable()` while also keeping as little redundant code as necessary. There might be a way to restructure this.

In general though, what do you think about the approach [here](https://github.com/atsamd-rs/atsamd/blob/16ff6e0afd3e5e01084149ad033ac04f5eb9fd82/hal/src/sercom/v2/spi.rs#L475-L488)? Each implementer of `Mode` maps to the corresponding PAC constants. To me, that makes perfect sense. You're mapping from types to values.

Would it help if the `bool` were named `XTALEN`, which matches the PAC field name? That way it's more clear that the type is mapping to the corresponding register fields.

glaeqen replied:
That'd work better IMO because associated constant provides actual data, not the condition value that is used to execute one code branch or the other somewhere else. However, it does not feel right that mode specfic configuration is partially done on a `Xosc::enable` level by using data stored in associated constants and partially in `fn configure/enable` level. It feels inconsistent. I like `mode` specific stuff to be hidden away in `fn configure/enable` although I admit you're absolutely right that order of execution matters which in that design is pure abstraction leakage and implementor must magically know to put `self.mode.enable/configure(&mut self.token)` in a proper spot in `Xosc::enable`. Other approach would assume that you get rid of `fn configure` altogether but then you end up with code repetition and nasty ifs. Hmm, tricky.

glaeqen replied:
In example you quoted, it's only data that is being abstracted away. In our example, as was already mentioned, we try to abstract both data and behaviour at the same time. And this does not seem totally right.

bradleyharden replied:
I understand your objection better now. And I agree. It probably is an anti-pattern to use an associated `bool` in a condition.

I think my goal will be to change the implementation so that `Mode<X>` *only* abstracts data and does not abstract behavior. I think that makes for the cleanest design. I think I actually performed a similar refactor in the `spi` module long ago. I used to have `configure` functions on the `Mode` trait, but I removed them in favor of the data-only approach.

Thanks for helping to clarify that.

bradleyharden replied:
I think the latest implementation resolves this issue.

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -311,37 +267,83 @@ impl<X: XoscNum> XoscToken<X> {
+    /// Private trait to restrict access to the `configure` function
+    pub trait Mode<X: XoscId>: Sealed {
+        const FROM_CRYSTAL: bool;
+        type Config;
```
I know you aim for 100% 0-sized structs everywhere but IMO `Config` should be an implementation detail of a downstream `Mode` trait implementor (internal field). Similarly to what I did in `dfll` module. Then you would have `fn configure(&mut Token);` function that is callable like `self.mode.enable(&mut self.token);`

Update: looking at the way you interact with modes, `Config` occupies memory space either way within `Xosc`, so there's really no drawbacks whatsoever - then again my point can be cynically reduced to just personal aesthetics appeal. Whatever path we take, I think we should unify the way modes are managed between `xosc` and `dfll`.

bradleyharden replied:
The purpose of the refactor wasn't to change the size of the struct. It was to remove the redundant type parameter when specifying `CrystalMode`. See the comment above. I was planning to do the same for DPLL. Actually, I think my write-up on the main PR talked about this issue with the DPLL. Did you see that?

glaeqen replied:
I've seen it and stopped half way because it was a lot to process at once. I'll come back to it

glaeqen commented on hal/src/thumbv7em/clock/v2/osculp32k.rs:
```diff
@@ -103,37 +93,19 @@ impl OscUlp32kToken {
 /// - An output state of a 1 kHz signal (active/inactive)
-pub struct OscUlp32k<X, Y>
-where
-    X: Output32k,
```
The same here, multiplexer switches must be reflected in types somehow. Otherwise it will be unsound.

glaeqen commented on hal/src/thumbv7em/clock/v2/rtcosc.rs:
```diff
@@ -0,0 +1,207 @@
+    fn enable_rtc_osc(&mut self, freq: DynRtcFreq) {
+        use DynRtcFreq::*;
+        match freq {
+            Freq1k => self.0.token.activate_1k(true),
```
As we spoke, 32k switch is also relevant for Gclks so this is not going to work (comment FTR)

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -154,19 +153,21 @@ pub struct Tokens {
     pub apbs: apb::ApbClks,
     /// Construction token for [`dpll::Dpll0`]
-    pub dpll0: dpll::DpllToken<dpll::marker::Dpll0>,
+    pub dpll0: dpll::DpllToken<dpll::DpllId0>,
```
I think it make sense for `*Id` types to have `Id` suffix at the actual end. `Dpll0Id` > `DpllId0`. @vcchtjader agreed with me on that matter but maybe you have a different take on this.

bradleyharden replied:
Yeah, I was leaning that way as well. The existing approach was the simplest to implement with `seq`. I'll have to figure out how to change it.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -221,25 +223,67 @@ pub fn retrieve_clocks(
-pub trait SourceMarker: crate::typelevel::Sealed {}
+/// For example, `Enabled<Gclk5<DfllId>>` would implement
+/// `Driver<Source = GclkId5>`.
+pub trait Driver: Sealed {
```
After taking some time I think now understand fully how this change is supposed to work. You moved the duplicated layering of `Gclk/Dpll/PclkSource` and `Gclk/Dpll/PclkSourceMarker` by removing and squashing the former into the `Source` (renamed into `Driver`) and removing the associated type constraint. Matching against the correct source is done within `impl` blocks by using `*Id` types. Earlier the matching was redundant. Very nice job! I got too narrow minded to pick such a thing up.

bradleyharden replied:
Precisely. I didn't recognize it either until I read that forum post and started thinking about redundant trait bounds in several different places.

Do we like `Driver`? Or should we go back to `Source`? If we go back to `Source`, I think it should be
```rust
pub trait Source: Sealed {
    type Id;
    fn freq(&self) -> Hertz;
}
```

I don't have particularly strong feelings. If I can try to restate my previous post briefly, I think I was aiming for:
> I **am** a driver. I drive other clocks. So I implement `Driver`

vs.

> I **receive** a clock source. So I look up the `Source` of my `Driver`.

I'm not sure if that distinction helps or not. I thought it did, but I think `Source` and `Id` would also work well.

glaeqen replied:
`Source` and `Id` feels better for me but that's aesthetics.

bradleyharden replied:
I think I'll switch to `Source` and `Id`. Now that there are fewer traits using the word `Source`, it feels less overloaded to me. I also like that `Id` clearly indicates that the type should be one of the `Id` types.

glaeqen commented on hal/src/thumbv7em/clock/v2/dpll.rs:
```diff
@@ -31,109 +31,109 @@ use typenum::U0;
-pub trait DpllSourceXosc32k: DpllSource {}
+pub trait DpllSourceXosc32k: Driver {}
+
+impl<M, N> DpllSourceXosc32k for Enabled<Xosc32k<M>, N>
```
Following the path of `Driver` idea, it should be possible to remove and replace this pattern with subhierarchy of `*Id` traits

```rust
pub trait DpllXosc32kBasedSourceId: DpllSourceId {}
pub trait DpllXoscBasedSourceId: DpllSourceId {}
```
and then implement it on corresponding `*Id` types. Then narrow down `T: DpllSourceId` clause of `Xosc32kDriven` and `XoscDriven` specialisations of `Dpll`  to `T: DpllXosc32kBasedSourceId` and `T: DpllXoscBasedSourceId` respectively. That should work I believe.

glaeqen replied:
I realized `DpllXosc32kBasedSourceId` can be just replaced with a specialisation for a concrete `Xosc32kId` type. The other one needs to stay to cover both `Xosc0Id` and `Xosc1Id`

bradleyharden replied:
I haven't gotten to the `dpll` module yet. But you hint at one of the issues I have with the `osculp32k` and `xosc32k` modules. We have a single type that represents multiple clocks. Right now, the `Driver` trait expects that each type has exactly one `Source` type. But `OscUlp32k<Active1k, Active32k>` could conceivably implement `Driver` twice, with two different `Source` types. It breaks the consistency, which is why I've been trying so hard to find a different way to phrase it.

glaeqen commented on hal/src/thumbv7em/clock/v2/gclkio.rs:
```diff
@@ -267,60 +278,39 @@ impl<G: GclkNum> GclkOutToken<G> {
+    use super::{Counter, Driver, Enabled, Gclk, GclkId, GclkSourceId, NotGclkInId};
 
-pub(crate) use private::GclkOutSource as PrivateGclkOutSource;
+    type EnabledGclk<T> = Enabled<
```
This feels like a _type hack_ and feels for me out of place. Isn't `AsEnabledGclk` redundant if we are matching on `Driver<Source = G>` where `G: GclkId`? Am I missing something?

glaeqen replied:
Right, it cannot be `GclkInId`.. I wonder if there's a simplier way.

bradleyharden replied:
Well, we could skip `Driver` here and directly require `Enabled<Gclk<G, S>, N>`. I believe that was the approach taken by the earliest implementation of `GclkOut`. I think you didn't like it though, because it was inconsistent with the rest of the `clock` module.

We could also go back to your approach that defined redundant `enable_gclk_out` functions on a `private` trait. That works too. I was just kind of experimenting with this approach.

I agree that it feels a little hacky. The least hacky approach is probably `Enabled<Gclk<G, S>, N>`.

bradleyharden replied:
I think this would compile. Looks pretty clean to me.

```rust
impl<G, I> GclkOut<G, I>
where
    G: GclkId,
    I: GclkIo<G>,
{
    pub fn enable<S, N>(
        token: GclkOutToken<G>,
        pin: impl AnyPin<Id = I>,
        mut gclk: Enabled<Gclk<G, S>, N>,
        polarity: bool,
    ) -> (GclkOut<G, I>, Enabled<Gclk<G, S>, N::Inc>)
    where
        S: NotGclkIn,
        N: Increment,
    {
        let freq = gclk.freq();
        let pin = pin.into().into_alternate();
        gclk.enable_gclk_out(polarity);
        let gclk_out = GclkOut { token, freq, pin };
        (gclk_out, gclk.inc())
    }
}
```

glaeqen replied:
Yeah, maybe we should operate directly on `Enabled<Gclk>`. Lesser evil. `GclkOut` has it's quirks so it's probably always going to be awkward to model within our design and conventions.

bradleyharden replied:
I agree. I would also prefer the abstract, `D: Driver` approach, but I don't think there is any way to do in cleanly. This is probably the cleanest approach. I'll make the switch.

glaeqen commented on hal/src/thumbv7em/clock/v2/xosc.rs:
```diff
@@ -422,9 +424,26 @@ where
+    /// Returns the enabled Xosc
+    #[inline]
+    pub fn enable(mut self) -> Enabled<Xosc<X, M>, U0> {
+        self.token.reset();
```
HW reset is a nice touch as it will harden this API against some misconfigured settings. Was lack of reset making an `Xosc` abstraction *very* unsound? I think you've mentioned that it did.

bradleyharden replied:
I don't think we necessarily need a reset if we set *all* settings in the call to `.enable()`. But I don't think the existing implementation did that.

I think the only way to trigger a problem in this specific case would be to `.enable()` `CrystalMode`, `.disable()`, and then re-`.enable()` in `ClockMode`. I don't think anyone would ever actually do that. But it's something we should look out for.

glaeqen replied:
> I think the only way to trigger a problem in this specific case would be to .enable() CrystalMode, .disable(), and then re-.enable() in ClockMode. I don't think anyone would ever actually do that. But it's something we should look out for.

Why would it cause an issue? Source is changed and in clock mode, settings set in `CrystalMode::configure` are irrelevant (right?). I'm missing something probably. It's not like it matters but I'm just curious.

bradleyharden replied:
Hmm, maybe it wouldn't have been an issue. I'm not sure. The previous implementation would set the crystal current, amplitude loop control and low buf gain *only* in crystal mode. If those settings have any effect on clock mode, they would not be reset when going from crystal mode to clock mode.

However, maybe those settings are completely ignored in clock mode. if that's the case, it probably doesn't matter. But I think I would still feel better knowing that we always start from a known state. Do you like that idea?

glaeqen replied:
I'm fine with it

vcchtjader replied:
This seems resolved?

I have no objection that we would like to remove possible unwanted statefulness and/or disconnects between abstraction model and hardware due to residual hardware configurations.

glaeqen reviewed (commented):
There's plenty of comments of mine all over the place but I think your suggestions are usually excellent. One thing I didn't mention in comments were docs inconsistencies but these can be dealt with later when this rework gets to the stable phase.

bradleyharden commented:
I'm adding my question from Matrix here.

Here's another broad question for both of you.

In the spi and uart modules, we tried to maintain a particular pattern when dealing with the Config and Spi/Uart structs. For a given setting, the Config struct has three methods:

.get_setting(&self)
.set_setting(&mut self, ...)
.setting(self, ...) -> Self
The versions taking self by value are the primary API, because they can be chained together and finalized with a call to .enable(). However, we included the .get_ and .set_ versions to allow the following pattern.

The Spi and Uart structs implement AsRef<Config>, so you can always get read access to those properties. This approach means we don't have to repeat those methods on both Config and Spi. Furthermore, the Spi and Uart structs have a .reconfigure<F>(&mut self, F) method that takes a closure F: FnOnce(&mut Config). The .reconfigure function temporarily disables the peripheral, runs the closure, and re-enables it. Again, we're able to provide access to the methods on Config without repeating them on the Spi/Uart structs.

Note that some settings do break this pattern. In particular, any setting that changes a type parameter must take self by value, so it can't have a .set_ version or be used with .reconfigure.

I was wondering if we might want to establish a similar convention for the clock module, although I think it would have to be slightly different. Obviously functions taking self by value will still be the primary API. But we could also selectively implement .set_ versions as well. Then, we could implement Deref and DerefMut for Enabled<T, N>. Any functions with a .set_ method would be available, but any methods with a self-by-value-only API would not.

Do you think that would work? It would save us repeated definitions of functions on Enabled<T, N>.

Maybe there's a problem with that approach I'm not seeing?

bradleyharden commented:
Also, let me reference the main PR, for posterity. [#450](https://github.com/atsamd-rs/atsamd/pull/450)

vccggorski commented:
> I'm adding my question from Matrix here.
> 
> Here's another broad question for both of you.
> 
> In the `spi` and `uart` modules, we tried to maintain a particular pattern when dealing with the `Config` and `Spi`/`Uart` structs. For a given setting, the `Config` struct has three methods:
> 
> * `.get_setting(&self)`
> * `.set_setting(&mut self, ...)`
> * `.setting(self, ...) -> Self`
> 
> The versions taking `self` by value are the primary API, because they can be chained together and finalized with a call to `.enable()`. However, we included the `.get_` and `.set_` versions to allow the following pattern.
> 
> The `Spi` and `Uart` structs implement `AsRef<Config>`, so you can always get read access to those properties. This approach means we don't have to repeat those methods on both `Config` and `Spi`. Furthermore, the `Spi` and `Uart` structs have a `.reconfigure<F>(&mut self, F)` method that takes a closure `F: FnOnce(&mut Config)`. The `.reconfigure` function temporarily disables the peripheral, runs the closure, and re-enables it. Again, we're able to provide access to the methods on `Config` without repeating them on the `Spi`/`Uart` structs.
> 
> Note that some settings do break this pattern. In particular, any setting that changes a type parameter **must** take `self` by value, so it can't have a `.set_` version or be used with `.reconfigure`.
> 
> I was wondering if we might want to establish a similar convention for the `clock` module, although I think it would have to be slightly different. Obviously functions taking `self` by value will still be the primary API. But we could also selectively implement `.set_` versions as well. Then, we could implement `Deref` and `DerefMut` for `Enabled<T, N>`. Any functions with a `.set_` method would be available, but any methods with a `self`-by-value-only API would not.
> 
> Do you think that would work? It would save us repeated definitions of functions on `Enabled<T, N>`.

Also copied from Matrix

First thought that pops up in my head is that clock abstractions, once Enabled, do not really have that many settings (as it would break their soundness) to benefit from Derefing into underlying type. Unless I missed your point, sorry

bradleyharden commented:
Yeah, you're probably right. It just *feels* to me like `Enabled` should implement `Deref` in some form, since it's such a thin wrapper. But then again, you could also argue that the entire purpose of `Enabled` is as a wrapper that removes access to the underlying functions.

bradleyharden committed:
clock/v2: Update 1 kHz, 32 kHz and RTC clocks

Refactor the slow osccilators to mimic the structure in the rest of the
`clock` module. Rather than use extra type parameters to represent the
enabled or disabled state of the 1 kHz and 32 kHz outputs, instead
create a `Base` clock that can only be used to enable downstream 1 kHz
and 32 kHz clocks. With this approach, the information is encoded in the
type parameter `N` of `Enabled<Clock, N>`, rather than dedicated type
parameters `X`` and `Y`. This approach improves consistency with the
rest of the `clock` module.

bradleyharden committed:
Xosc: Update `xosc::Mode` and `xosc::CrystalCurrent`

Use the `Mode` trait to map to the `XOut` type, allowing it to be
removed from the `CrystalConfig`. Merge `CrystalConfig` and
`CrystalMode`. Now `CrystalMode` is no longer generic. Remove any
abstraction of behavior from the `Mode` trait. Instead, have it
represent an interface to the underlying configuration data. For
`ClockMode`, the underlying values are the reset values for the
corresponding fields.

Remove the implementation of `From<Hertz>` for `CrystalMode`, which also
removes any possibility of failure. Instead, for the user to choose a
`CrystalCurrent` manually, when calling `from_crystal`. Add a `Zero`
variant to `CrystalCurrent` to represent the default values of `IMULT`
and `IPTAT`.

vccggorski commented:
> Yeah, you're probably right. It just _feels_ to me like `Enabled` should implement `Deref` in some form, since it's such a thin wrapper. But then again, you could also argue that the entire purpose of `Enabled` is as a wrapper that removes access to the underlying functions.

Implementing `Deref` on `Enabled` is probably harmless as, well, it won't allow mutating the underlying clock through immutable reference. `DerefMut` is definitely not acceptable. My point was that you don't really have that much to deref through it after all, as most of the existing API focuses on mutating the clock state.

glaeqen commented on hal/src/thumbv7em/clock/v2/osculp32k.rs:
```diff
@@ -1,129 +1,125 @@
-    pub fn set_calibration(mut self, calib: u8) -> Self {
-        self.0.token.set_calib(calib);
-        self
+    pub fn write_lock(mut self) {
```
In my opinion we should get rid of  `write_lock` API because it is unsound in case of non-POR based reset. Or at least maybe mark it unsafe? What do you think?

bradleyharden replied:
The write lock *is* POR based in this case. From the documentation:

> It is also possible to lock the XOSC32K configuration by setting the Write Lock bit in the 32 kHz
External Crystal Oscillator Control register (XOSC32K.WRTLOCK=1). If set, the XOSC32K
configuration is locked until a Power-On Reset (POR) is detected.

Is that not your question?

vcchtjader replied:
I think what @vccggorski means is that this lock is hard to model for any reset/change of application where the lock state is not dynamically picked up.

This might work just fine in a future Dyn-clocktree setup but as soon as the “Locked<X>” type is gone there is a disconnect between the abstraction and hardware.

Example:
Bootloader sets up a minimal clock tree and locks a few clocks here and there. Bootloader hands off to main application which is unaware of the exact clock tree setup. This would be fine if main application never had to touch the clock tree.

But bootloaders in contrast to main applications tends to be small and simple (with matching clocking needs) and thus the odds that modifications to clocking is required is quite high.

Trying to reset and reconfigure the locked parts of clock tree would fail and in the case it just _ignores_ the new setup (not sure about the implications of writing to write locked registers) there would be a huge disconnect between clock abstraction and actual hardware.

Maybe the simplest for now is to keep the method available but `unsafe` and let any users ensure themselves that it works for their specific needs?

bradleyharden replied:
I think the bootloader problem is a separate issue. Right now, we only provide an API to create a representation of the clock tree at POR. If that representation is not accurate for any reason, then all bets are off. It's essentially UB. Without knowing the state of the clock tree, we can't be sure that *anything* will work correctly.

The solution is to provide an `unsafe` way for users to create their own clock tree initialization, without writing any registers. I think that is what initially led @vccggorski to move all hardware writes to the `.enable()` functions.

Alternatively, if the state of the clock tree cannot be known until run-time, then our existing, compile-time API is insufficient. The only choice is to wait for a future `Dyn` API. I don't see any way around that.

Maybe you're suggesting that parts of the clock tree could be known statically while other parts could only be known dynamically? Even if that's true, you could still create an accurate representation of the static parts.

However, if you can be sure that your clock tree is accurate (which I claim **must** be true, or you essentially have UB), then I think the `write_lock` function accurately models the write lock.

glaeqen commented on hal/src/thumbv7em/clock/v2/rtcosc.rs:
```diff
@@ -66,142 +70,62 @@ impl RtcOscToken {
 
+impl RtcSourceId for OscUlp1kId {
+    const DYN: DynRtcSourceId = DynRtcSourceId::OscUlp1k;
+    const FREQ: Hertz = Hertz(1024);
```
I think that `DynRtcSourceId` should be able to convert itself `RTCSEL_A`.
Also, I'm not sure if you're planning to use `FREQ` in a const context later but maybe it can be generalized by using `DynRtcSourceId`. Information is somewhat duplicated :thinking:. Maybe not through `From` trait implementation but this is just an example.

```diff
diff --git a/hal/src/thumbv7em/clock/v2/rtcosc.rs b/hal/src/thumbv7em/clock/v2/rtcosc.rs
index 799babc5c..a0a40f64b 100644
--- a/hal/src/thumbv7em/clock/v2/rtcosc.rs
+++ b/hal/src/thumbv7em/clock/v2/rtcosc.rs
@@ -30,15 +30,7 @@ impl Registers {
 
     #[inline]
     fn set_source_freq(&mut self, source: DynRtcSourceId) {
-        use DynRtcSourceId::*;
-        use RTCSEL_A::*;
-        let variant = match source {
-            OscUlp1k => ULP1K,
-            OscUlp32k => ULP32K,
-            Xosc1k => XOSC1K,
-            Xosc32k => XOSC32K,
-        };
-        self.rtcctrl().write(|w| w.rtcsel().variant(variant));
+        self.rtcctrl().write(|w| w.rtcsel().variant(source.into()));
     }
 
     fn reset_source_freq(&mut self) {
@@ -76,26 +68,47 @@ pub enum DynRtcSourceId {
     Xosc32k,
 }
 
+impl From<DynRtcSourceId> for Hertz {
+    fn from(v: DynRtcSourceId) -> Self {
+        use DynRtcSourceId::*;
+        match v {
+            Xosc1k | OscUlp1k => Hertz(1024),
+            Xosc32k | OscUlp32k => Hertz(32_768),
+        }
+    }
+}
+
+impl From<DynRtcSourceId> for RTCSEL_A {
+    fn from(v: DynRtcSourceId) -> Self {
+        use DynRtcSourceId::*;
+        use RTCSEL_A::*;
+        match v {
+            OscUlp1k => ULP1K,
+            OscUlp32k => ULP32K,
+            Xosc1k => XOSC1K,
+            Xosc32k => XOSC32K,
+        }
+    }
+}
+
 pub trait RtcSourceId {
     const DYN: DynRtcSourceId;
-    const FREQ: Hertz;
 }
 
 impl RtcSourceId for OscUlp1kId {
     const DYN: DynRtcSourceId = DynRtcSourceId::OscUlp1k;
-    const FREQ: Hertz = Hertz(1024);
 }
+
 impl RtcSourceId for OscUlp32kId {
     const DYN: DynRtcSourceId = DynRtcSourceId::OscUlp32k;
-    const FREQ: Hertz = Hertz(32_768);
 }
+
 impl RtcSourceId for Xosc1kId {
     const DYN: DynRtcSourceId = DynRtcSourceId::Xosc1k;
-    const FREQ: Hertz = Hertz(1024);
 }
+
 impl RtcSourceId for Xosc32kId {
     const DYN: DynRtcSourceId = DynRtcSourceId::Xosc32k;
-    const FREQ: Hertz = Hertz(32_768);
 }
 
 //==============================================================================
```

bradleyharden replied:
Now that you mention it, I think `FREQ` was left over from a previous approach to the implementation. I can probably just remove it.

bradleyharden replied:
Actually, I might have been planning ahead. We can eventually refactor the `Rtc` to take ownership of the `RtcOsc` (just like other peripherals will take ownership of the corresponding `Pclk`). In that case, the `Rtc` will need to get the frequency. Adding the frequency as a constant on `RtcSourceId` seemed like the simplest way to do that.

Do you have any other ideas?

glaeqen reviewed (commented):
In general I'm really happy with where this is going :) Do you want to add anything more or can I integrate it into the main PR? We can always add more in the following PRs.

bradleyharden committed:
clock/v2: Refactor `dpll` module

Refactor the `dpll` module to remove the `*Driven` types and use `*Id`
types for the second type parameter. In doing so, reimplement
`PclkSourceId` and `DynPclkSourceId` in terms of `GclkId` and a new
`DynGclkId`. Make the `Pclk` types effectively aliases for the `Gclk`
types, since a `Pclk` can only be sourced by a `Gclk`. Remove the
`dpll::SrcMode` trait and move elements of it to the `DpllSourceId`
trait. Refactor the `Dpll` methods to be compatible with the new trait
structure. Also update some documentation throughout.

bradleyharden committed:
clock/v2: Rename `Driver`/`Source` to `Source`/`Id`

bradleyharden committed:
clock/v2: Refactor the `dfll` module

Refactor the `dfll` module to be more consistent with conventions
throughtout the reset of the `clock` module.

bradleyharden committed:
rtcosc: Add documentation for the `rtcosc` module

bradleyharden committed:
gclk: Remove `NotGclk1` traits

These traits are not necessary, because unique tokens already make it
impossible to construct a `Gclk1` sourced by `Gclk1`.

bradleyharden committed:
dpll: Rename `Dpll::from_xosc` for each `Xosc`

With the existing structure, the compiler could not infer the correct
types. Distinct function names is the most straightforward solution.

bradleyharden committed:
clock/2: Overhaul preset clock configurations

Reimplement the preset clock configurations as real functions and
structs, rather than macros. The result has clear benefits, because the
preset configurations are now checked at compile-time for the HAL,
unlike the macro versions, which are not checked until compile-time for
users.

To implement these changes, modify the way `PclkId` types are generated
and used to implemen the `pclk::Tokens` struct. The new approach allows
better reuse of the `PclkId` types throughout the HAL, including the new
`clock::configurations` module.

bradleyharden commented:
Accidentally pushed to the wrong branch. Sorry. Trying to undo it now

bradleyharden commented:
Can you "unmerge" this? Is that possible? I already forced your branch back to its original position.

bradleyharden commented:
Ok, since we can't seem to recover, I'm going to re-merge this and create a new PR with commits since then. Does that work?
